### ISSUE TITLE
suspendCoroutine's continuation.resume() work with Result.failure(Throwable) will throw exception

### ISSUE THREAD

**Issue by MattHPO** [Kotlin/kotlinx.coroutines#2218]
Hello! I am Matt from Taiwan. Engaged in Android development.
Recently I migrated my app from Kotlin v1.3.72 and Coroutine v1.3.7 to Kotlin v1.4.0 and coroutine v1.3.9, then encounter a problem... 
Did I overlook something in the migration?
(I didn't modify my code when migration)

When the app launch, the below code snippet will be invoked.
```
viewModelScope.launch {
  val someObject = withContext(bgDispatcher) {
    var result: SomeObject? = null
    if (result == null) {
      result = getSomeObjectByAAA(intent).getOrNull()
    }
    if (result == null) {
      result = getSomeObjectByBBB(intent, activity).getOrNull()
    }
    if (result == null) {
      result = getSomeObjectByCCC(intent, activity).getOrNull()
    }
    if (result == null) {
      result = intent.data?.let { SomeObject(it.toString()) }
    }
    result
  }

```
in getSomeObjectByBBB & getSomeObjectByCCC
There have
```
suspendCoroutine { continuation ->
	someFun.getSomeObject()
		.addOnFailureListener { continuation.resume(Result.failure(Exception())) }
		.addOnSuccessListener { continuation.resume(Result.success()) }
}
```

And the `resume` fun will throw the Exception...

What I expect is that the `Exception` in `Result.failure()` just like a value, and can use `Result.getOrNull()` to obtain `null` when Result.isFailure...

**Comment by shmuelr** [comment]
This looks very similar to the bug I filed
https://github.com/Kotlin/kotlinx.coroutines/issues/2209
which might have something to do with the new inference engine in 1.4

@MattHPO can you try compiling with '-XXLanguage:-NewInference' so see if that solves the issue here?
Thanks!

**Comment by qwwdfsad** [comment]
Hi, could you please attach the stacktrace and/or self-contained reproducer that I can copy-paste and run?

**Comment by MiSikora** [comment]
@qwwdfsad I encountered it yesterday. Here is a sample that shows this.

```kotlin
suspend fun main() {
    val failure = fail()
    println(failure)
}

suspend fun fail() = suspendCoroutine<Result<String>> {
    Executors.newSingleThreadExecutor().execute {
        it.resume(Result.failure(RuntimeException()))
    }
}
```

```
Exception in thread "main" java.lang.RuntimeException
	at MainKt$fail$2$1.run(main.kt:12)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

It works with Kotlin `1.3.72`, fails with `1.4.0`, and works again with `1.4.20-dev-3898-14`. When `fail()` is changed to be an `inline` function it works again even with `1.4.0`. You can find more info [here](https://kotlinlang.slack.com/archives/C1CFAFJSK/p1599507442145100).

**Comment by qwwdfsad** [comment]
Thanks for the details!

This is a duplicate of #2181 and, unfortunately, a compiler bug [KT-41429](https://youtrack.jetbrains.com/issue/KT-41429) related to returning `Result` class from suspend functions that will be fixed in Kotlin 1.4.20.
